Crash Report: java - v1.5.0 (android)

## Crash Report

**Version:** 1.5.0
**Platform:** android
**Device:** android localhost


### Crash Log
```
=== JAVA CRASH REPORT ===
Time: 2026-02-15 08:26:34
Thread: QtBluetoothLEHandlerThread
Exception: java.lang.RuntimeException
Message: android.os.DeadObjectException

=== STACK TRACE ===
java.lang.RuntimeException: android.os.DeadObjectException
	at android.bluetooth.BluetoothGatt.writeCharacteristic(BluetoothGatt.java:1881)
	at org.qtproject.qt.android.bluetooth.QtBluetoothLE.executeWriteJob(QtBluetoothLE.java:1699)
	at org.qtproject.qt.android.bluetooth.QtBluetoothLE.performNextIO(QtBluetoothLE.java:1554)
	at org.qtproject.qt.android.bluetooth.QtBluetoothLE.access$1600(QtBluetoothLE.java:42)
	at org.qtproject.qt.android.bluetooth.QtBluetoothLE$3.run(QtBluetoothLE.java:1489)
	at android.os.Handler.handleCallback(Handler.java:995)
	at android.os.Handler.dispatchMessage(Handler.java:103)
	at android.os.Looper.loopOnce(Looper.java:273)
	at android.os.Looper.loop(Looper.java:363)
	at android.os.HandlerThread.run(HandlerThread.java:85)
Caused by: android.os.DeadObjectException
	at android.os.BinderProxy.transactNative(Native Method)
	at android.os.BinderProxy.transact(BinderProxy.java:661)
	at android.bluetooth.IBluetoothGatt$Stub$Proxy.writeCharacteristic(IBluetoothGatt.java:1211)
	at android.bluetooth.BluetoothGatt.writeCharacteristic(BluetoothGatt.java:1859)
	... 9 more

=== DEVICE INFO ===
Android: 16 (API 36)
Device: samsung SM-X115

```

### Debug Log (last lines)
```
[169770.682] DEBUG Resending steam settings to keep heater on
[169770.683] DEBUG applySteamSettings: sending 160 °C (reason: "keepSteamHeaterOn=true" , phase: "4" , configuredTemp: 160 )
[169770.733] DEBUG DE1Device: Writing to "0000a00b" data: "00a02455c83c245d00"
[169770.823] DEBUG DE1Device: Write confirmed to "0000a00b" data: "00a02455c83c245d00"
[169770.823] DEBUG DE1Device: Writing to "0000a006" data: "04803828d2000000000000000000000000000000"
[169770.972] DEBUG DE1Device: Write confirmed to "0000a006" data: "04803828d2000000000000000000000000000000"
[169770.973] DEBUG DE1Device: Writing to "0000a006" data: "04803848f6090000000000000000000000000000"
[169771.069] DEBUG DE1Device: Write confirmed to "0000a006" data: "04803848f6090000000000000000000000000000"
[169800.524] DEBUG DE1Device: Writing to "0000a006" data: "0480385401000000000000000000000000000000"
[169800.612] DEBUG DE1Device: Write confirmed to "0000a006" data: "0480385401000000000000000000000000000000"
[169818.483] DEBUG Sleep countdown: normal=9, stayAwake=37
[169830.682] DEBUG Resending steam settings to keep heater on
[169830.683] DEBUG applySteamSettings: sending 160 °C (reason: "keepSteamHeaterOn=true" , phase: "4" , configuredTemp: 160 )
[169830.732] DEBUG DE1Device: Writing to "0000a00b" data: "00a02455c83c245d00"
[169830.838] DEBUG DE1Device: Write confirmed to "0000a00b" data: "00a02455c83c245d00"
[169830.839] DEBUG DE1Device: Writing to "0000a006" data: "04803828d2000000000000000000000000000000"
[169830.935] DEBUG DE1Device: Write confirmed to "0000a006" data: "04803828d2000000000000000000000000000000"
[169830.936] DEBUG DE1Device: Writing to "0000a006" data: "04803848f6090000000000000000000000000000"
[169831.033] DEBUG DE1Device: Write confirmed to "0000a006" data: "04803848f6090000000000000000000000000000"
[169860.523] DEBUG DE1Device: Writing to "0000a006" data: "0480385401000000000000000000000000000000"
[169860.625] DEBUG DE1Device: Write confirmed to "0000a006" data: "0480385401000000000000000000000000000000"
[169878.495] DEBUG Sleep countdown: normal=8, stayAwake=36
[169890.682] DEBUG Resending steam settings to keep heater on
[169890.683] DEBUG applySteamSettings: sending 160 °C (reason: "keepSteamHeaterOn=true" , phase: "4" , configuredTemp: 160 )
[169890.734] DEBUG DE1Device: Writing to "0000a00b" data: "00a02455c83c245d00"
[169890.804] DEBUG DE1Device: Write confirmed to "0000a00b" data: "00a02455c83c245d00"
[169890.805] DEBUG DE1Device: Writing to "0000a006" data: "04803828d2000000000000000000000000000000"
[169890.901] DEBUG DE1Device: Write confirmed to "0000a006" data: "04803828d2000000000000000000000000000000"
[169890.902] DEBUG DE1Device: Writing to "0000a006" data: "04803848f6090000000000000000000000000000"
[169890.997] DEBUG DE1Device: Write confirmed to "0000a006" data: "04803848f6090000000000000000000000000000"
[169920.524] DEBUG DE1Device: Writing to "0000a006" data: "0480385401000000000000000000000000000000"
[169920.589] DEBUG DE1Device: Write confirmed to "0000a006" data: "0480385401000000000000000000000000000000"
[169938.473] DEBUG Sleep countdown: normal=7, stayAwake=35
[169950.684] DEBUG Resending steam settings to keep heater on
[169950.685] DEBUG applySteamSettings: sending 160 °C (reason: "keepSteamHeaterOn=true" , phase: "4" , configuredTemp: 160 )
[169950.750] DEBUG DE1Device: Writing to "0000a00b" data: "00a02455c83c245d00"
[169950.912] DEBUG DE1Device: Write confirmed to "0000a00b" data: "00a02455c83c245d00"
[169950.913] DEBUG DE1Device: Writing to "0000a006" data: "04803828d2000000000000000000000000000000"
[169951.009] DEBUG DE1Device: Write confirmed to "0000a006" data: "04803828d2000000000000000000000000000000"
[169951.009] DEBUG DE1Device: Writing to "0000a006" data: "04803848f6090000000000000000000000000000"
[169951.110] DEBUG DE1Device: Write confirmed to "0000a006" data: "04803848f6090000000000000000000000000000"
[169980.522] DEBUG DE1Device: Writing to "0000a006" data: "0480385401000000000000000000000000000000"
[169980.602] DEBUG DE1Device: Write confirmed to "0000a006" data: "0480385401000000000000000000000000000000"
[169998.472] DEBUG Sleep countdown: normal=6, stayAwake=34
[170010.682] DEBUG Resending steam settings to keep heater on
[170010.691] DEBUG applySteamSettings: sending 160 °C (reason: "keepSteamHeaterOn=true" , phase: "4" , configuredTemp: 160 )
[170010.740] DEBUG DE1Device: Writing to "0000a00b" data: "00a02455c83c245d00"

========== SESSION START: 2026-02-15T08:30:55 ==========
[   0.048] DEBUG App started - version 1.5.0
```

---
*Auto-reported by Decenza DE1 app*

--- a/android/src/io/github/kulitorum/decenza_de1/DecenzaActivity.java
+++ b/android/src/io/github/kulitorum/decenza_de1/DecenzaActivity.java
@@ -65,16 +65,28 @@ public class DecenzaActivity extends QtActivity {
         super.onDestroy();
     }
 
-    // Catches both DeadObjectException (the common case: the Bluetooth GATT
-    // binder/process died — toggled off, OEM power policy, driver hiccup) and
-    // its subclass DeadSystemException (system_server died — deep sleep, OEM
-    // power management). Qt's QtBluetoothLE.executeWriteJob doesn't wrap the
+    // Catches three closely-related dead-binder cases:
+    //   1. DeadObjectException — the common case: a remote Bluetooth binder
+    //      we were writing to died (BT toggled off, OEM power policy, driver
+    //      hiccup). Covered by `instanceof DeadObjectException`. (Issue #1227)
+    //   2. DeadSystemException — system_server died. Subclass of
+    //      DeadObjectException so the same `instanceof` covers it.
+    //   3. DeadSystemRuntimeException — the API 31+ unchecked variant of
+    //      case 2. **Despite the name, it does NOT extend DeadObjectException**
+    //      (it extends RuntimeException directly), and is typically thrown
+    //      without a DeadObjectException in its cause chain, so the
+    //      `instanceof` check misses it. We fall back to a class-name
+    //      substring match to cover it. (Was caught by the original
+    //      `contains("DeadSystem")` filter in PR #544 / issue #538.)
+    // Qt's QtBluetoothLE.executeWriteJob doesn't wrap the
     // BluetoothGatt.writeCharacteristic binder call in try/catch (verified
     // through Qt 6.11.1), so the RuntimeException it raises lands here.
-    // Issues #189 (v1.5.0) and #1227 (v1.7.3).
+    // Issues #189 (v1.5.0), #538 (v1.4.x DeadSystemRuntimeException),
+    // #1227 (v1.7.3 DeadObjectException).
     private static boolean isDeadObjectException(Throwable t) {
         while (t != null) {
-            if (t instanceof DeadObjectException) {
+            if (t instanceof DeadObjectException
+                    || t.getClass().getName().contains("DeadSystem")) {
                 return true;
             }
             t = t.getCause();

--- a/android/src/io/github/kulitorum/decenza_de1/DecenzaActivity.java
+++ b/android/src/io/github/kulitorum/decenza_de1/DecenzaActivity.java
@@ -2,6 +2,7 @@ package io.github.kulitorum.decenza_de1;
 
 import android.content.Intent;
 import android.os.Bundle;
+import android.os.DeadObjectException;
 import android.util.Log;
 
 import org.qtproject.qt.android.bindings.QtActivity;
@@ -64,9 +65,16 @@ public class DecenzaActivity extends QtActivity {
         super.onDestroy();
     }
 
-    private static boolean isDeadSystemException(Throwable t) {
+    // Catches both DeadObjectException (the common case: the Bluetooth GATT
+    // binder/process died — toggled off, OEM power policy, driver hiccup) and
+    // its subclass DeadSystemException (system_server died — deep sleep, OEM
+    // power management). Qt's QtBluetoothLE.executeWriteJob doesn't wrap the
+    // BluetoothGatt.writeCharacteristic binder call in try/catch (verified
+    // through Qt 6.11.1), so the RuntimeException it raises lands here.
+    // Issues #189 (v1.5.0) and #1227 (v1.7.3).
+    private static boolean isDeadObjectException(Throwable t) {
         while (t != null) {
-            if (t.getClass().getName().contains("DeadSystem")) {
+            if (t instanceof DeadObjectException) {
                 return true;
             }
             t = t.getCause();
@@ -78,16 +86,18 @@ public class DecenzaActivity extends QtActivity {
         defaultHandler = Thread.getDefaultUncaughtExceptionHandler();
 
         Thread.setDefaultUncaughtExceptionHandler((thread, throwable) -> {
-            // DeadSystemException means Android's Bluetooth system service died
-            // (deep sleep, OEM power management, system_server crash).
-            // Create a flag file so the C++ side can detect it and trigger BLE
-            // recovery (disconnect + reconnect). This fires on any thread — we
-            // don't filter by thread name since DeadSystemException only occurs
-            // when Android system services die, which always affects BLE.
-            // Don't crash the app — the rest of the app is fine, only BLE is dead.
-            if (isDeadSystemException(throwable)) {
-                Log.w(TAG, "DeadSystemException on thread " + thread.getName()
-                        + " — Bluetooth system died, signaling BLE recovery");
+            // A DeadObjectException means the remote Bluetooth binder we
+            // were writing to died (toggled off, OEM power-managed away,
+            // GATT proxy unbound) — or, in the DeadSystemException subclass
+            // case, system_server itself died. In all of these the BLE
+            // handler thread is gone but the UI thread and the rest of the
+            // app are fine, so we drop a flag file for the C++ side to
+            // observe and trigger reconnect (see main.cpp ble recovery
+            // timer), and explicitly do NOT call defaultHandler — the
+            // process must stay alive.
+            if (isDeadObjectException(throwable)) {
+                Log.w(TAG, "DeadObjectException on thread " + thread.getName()
+                        + " — BLE binder died, signaling BLE recovery");
                 try {
                     File flagFile = new File(getFilesDir(), "ble_dead_system");
                     flagFile.createNewFile();

--- a/src/main.cpp
+++ b/src/main.cpp
@@ -866,11 +866,12 @@ int main(int argc, char *argv[])
         "idleGc", "()V");
     idleGcTimer->start();
 
-    // BLE dead system recovery: Android's Bluetooth system service can die
-    // (DeadSystemException) during deep sleep or OEM power management.
-    // When this happens, the Qt BLE handler thread dies but the app keeps running.
-    // The Java crash handler writes a flag file; we check for it every 10 seconds
-    // and trigger BLE reconnection if found.
+    // BLE dead-binder recovery: when the Bluetooth GATT binder/process dies
+    // (toggled off, OEM power policy, GATT proxy unbound) Qt's BLE handler
+    // thread raises a DeadObjectException (or its DeadSystemException
+    // subclass if system_server itself died). The Java crash handler catches
+    // both, keeps the app alive, and writes a flag file; we poll for it
+    // every 10 s and trigger BLE reconnection if found. Issues #189, #1227.
     auto* bleRecoveryTimer = new QTimer();
     bleRecoveryTimer->setInterval(10000);
     QObject::connect(bleRecoveryTimer, &QTimer::timeout,
@@ -882,7 +883,7 @@ int main(int argc, char *argv[])
         if (!flagFile.exists())
             return;
 
-        qWarning() << "BLE recovery: DeadSystemException detected, triggering reconnect";
+        qWarning() << "BLE recovery: dead BLE binder detected, triggering reconnect";
         flagFile.remove();
 
         // The BLE handler thread is dead — Qt's QLowEnergyController won't emit

